### DISCUSSION TITLE
Problem with umqtt.simple

### DISCUSSION THREAD

**Discussion by Piootre** [micropython/micropython#13508] [ESP8266]
Hi, I have a problem with the check_msg() function in umqtt.simple.
I don't know why I'm getting an error with this function. Below is the code and error text.
Thank you very much for your tips. (I use esp8266)

code:
```python
import network
import time
import machine
import ubinascii
from umqtt.simple import MQTTClient

# WiFi
wifi_ssid = "ssid"
wifi_password = "password"

# MQTT
mqtt_broker = "broker.hivemq.com"
mqtt_port = 1883
mqtt_user = "user"
mqtt_password = "user123"
mqtt_topic = b"test/test"


def do_connect():
    sta_if = network.WLAN(network.STA_IF)
    if not sta_if.isconnected():
        print('connecting to network...')
        sta_if.active(True)
        sta_if.connect(wifi_ssid, wifi_password)
        
        while not sta_if.isconnected():
            print("Attempting to connect....")
            time.sleep(1)
            
        print('Connected! Network config:', sta_if.ifconfig())
        
def sub_cb(topic, msg):
    print(topic, msg)
    
def reset():
    print("Resetting...")
    time.sleep(5)
    machine.reset()

def connect_sub():
    print(f"Begin connection with MQTT Broker :: {mqtt_broker}")
    mqttClient = MQTTClient('esp8266', mqtt_broker,mqtt_port, mqtt_user,mqtt_password, keepalive=60)
    mqttClient.set_callback(sub_cb)
    mqttClient.connect()
    mqttClient.subscribe(mqtt_topic)
    print(f"Connected to MQTT  Broker :: {mqtt_broker}, Topic :: {mqtt_topic}")
    return mqttClient
            
try:
    do_connect()
    mqttClient = connect_sub()
    print(mqttClient.check_msg())
except OSError as e:
    print("Error: ",e)
    #reset()
while True:
    try:
        mqttClient.wait_msg()
    except OSError as e:
        print("Error in check_msg: ",e)
        reset()
```

Answer:

>>> %Run -c $EDITOR_CONTENT
Begin connection with MQTT Broker :: broker.hivemq.com
Connected to MQTT  Broker :: broker.hivemq.com, Topic :: b'test/test'
None
Error in check_msg:  -1
Resetting...

**Comment by robert-hh** [comment]
Please enclose the code in lines consisting of three backticks ```
At the fist backtick line you can append `python` to get syntax coloring.

**Reply by Piootre** [reply]
done

**Reply by nastiliano** [reply]
Hello,

I recommend you try the latest version (there are a lot of fixed bugs):

https://github.com/fizista/micropython-umqtt.simple2


Best regards.

**Reply by Piootre** [reply]
Now I have error 'MQTTException: 1' in  line: mqttClient.wait_msg()
```python
while True:
    try:
        mqttClient.wait_msg()
    except OSError as e:
        print("Error in check_msg: ",e)
        reset()
```
I found in the documentation that the TOPIC format is wrong but I don't understand why:

Different problems
Wrong topic format during subscription - you'll get OSError: [Errno 104] ECONNRESET in subscribe() or MQTTException: 1 in the wait_msg()/check_msg()

**Reply by nastiliano** [reply]
Mmmm... I'm not an expert at all. You will need the experts help.

But I'd try to quit:  "print(mqttClient.check_msg())" line

Try to catch the "msg "directly in the fuction
    def sub_cb(topic, msg):
        print(topic, msg)

Put it directly in a variable to use it after.

I think the topic is good. Try to put something simple, without "/", just "test"
Best regards.

**Reply by Piootre** [reply]
in while loop I change on "print('in while: ',mqttClient.wait_msg())"  executes once and returns a value of 144
next is the same error

**Comment by Piootre** [comment]
I don't know why, but the code below works. I tried changing my code to make it look the same, but my code still didn't work xD

Source: https://github.com/avseng/MQTT-RPI-ESP8266

Code:
```python
from umqtt.simple2 import MQTTClient
import network
from machine import Pin
from time import sleep


MQTT_SERVER = "broker.hivemq.com"
CLIENT_ID = ""
MQTT_TOPIC = b""
MQTT_PASS = ""
MQTT_SSID = ""

WIFI_SSID = ""
WIFI_PASSWORD = ""

#led = Pin(2, Pin.OUT)

# WiFi connectivity function.
def connectWIFI():
  wlan = network.WLAN(network.STA_IF)
  wlan.active(True)
  wlan.connect(WIFI_SSID,WIFI_PASSWORD)
  while not wlan.isconnected():
    pass

  print ("wifi connected")  
  print(wlan.ifconfig())
  
  
#MQTT callback service function
def sub_cb(topic, msg,r,t): 
  print(topic, msg)


try:
  #led.value(1)
    
  connectWIFI()
  c = MQTTClient(CLIENT_ID,MQTT_SERVER,1883,MQTT_SSID,MQTT_PASS)
  c.set_callback(sub_cb)
  c.connect()
  c.subscribe(MQTT_TOPIC)
  
  print("MQTT connected and ready to receive message")
  while True:
    c.wait_msg()
finally:
  c.disconnect()
```

**Comment by peterhinch** [comment]
@nastiliano Describing that as the latest version is incorrect. It is not a bugfix release, rather it is a user written library and is not officially supported. The current officially supported code is [here](https://github.com/micropython/micropython-lib/tree/master/micropython/umqtt.simple).

@Piootre The fault is occurring [here](https://github.com/micropython/micropython-lib/blob/7cdf70881519c73667efbc4a61a04d9c1a49babb/micropython/umqtt.simple/umqtt/simple.py#L183) indicating that the broker is returning an invalid response. I would check whether you have achieved a connection by doing a publication and testing that it can be received (e.g. by using `mosquitto-sub` on a PC).
